File: backend/scripts/check_coverage.py
#!/usr/bin/env python3
"""
Phase 0 #7: coverage ratchet gate for the backend test suite.

Why a custom script (instead of ``--cov-fail-under=X``):

*   Makes the baseline explicit in the repo (``.coverage-baseline``) so
    bumping the floor is a reviewable commit rather than a change nobody
    notices.
*   Gives clear ratchet semantics: "never drop below the committed baseline
    (minus a tiny tolerance)" — the number can be raised whenever a PR lands
    genuinely new tests.
*   Works with either ``coverage.xml`` or ``.coverage`` (SQLite) emitted by
    ``coverage.py``/``pytest-cov``.
*   Prints the gap to the aspirational target so CI readers see both the
    hard floor and the north star.

Usage::

    python backend/scripts/check_coverage.py                   # default paths
    python backend/scripts/check_coverage.py --coverage path.xml --baseline 0.40
    python backend/scripts/check_coverage.py --write-baseline  # ratchet up
"""
from __future__ import annotations

import argparse
import sys
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Optional

HERE = Path(__file__).resolve().parent
BACKEND_ROOT = HERE.parent
REPO_ROOT = BACKEND_ROOT.parent

# Defaults mirror `backend/.coveragerc` + layout.
DEFAULT_COVERAGE_XML = BACKEND_ROOT / "coverage.xml"
DEFAULT_BASELINE_FILE = BACKEND_ROOT / ".coverage-baseline"
DEFAULT_TOLERANCE_PP = 0.5  # percentage-points — absorbs rounding / minor flake.
DEFAULT_TARGET_PERCENT = 60.0  # audit plan aspirational backend target.


def _read_baseline(path: Path) -> float:
    if not path.is_file():
        raise SystemExit(f"[check_coverage] baseline file not found: {path}")
    raw = path.read_text(encoding="utf-8").strip()
    if not raw:
        raise SystemExit(f"[check_coverage] baseline file is empty: {path}")
    try:
        return float(raw)
    except ValueError as exc:
        raise SystemExit(
            f"[check_coverage] baseline file must contain a single float "
            f"(got {raw!r}): {path}"
        ) from exc


def _measure_from_xml(xml_path: Path) -> float:
    """Return overall coverage as a percentage (0-100) from Cobertura XML."""
    if not xml_path.is_file():
        raise SystemExit(
            f"[check_coverage] coverage report not found: {xml_path}\n"
            "Run `pytest --cov=backend/app --cov-report=xml:backend/coverage.xml` first."
        )
    tree = ET.parse(xml_path)
    root = tree.getroot()
    rate = root.get("line-rate")
    if rate is None:
        raise SystemExit(
            f"[check_coverage] 'line-rate' attribute missing from {xml_path}"
        )
    try:
        return float(rate) * 100.0
    except ValueError as exc:
        raise SystemExit(
            f"[check_coverage] malformed line-rate={rate!r} in {xml_path}"
        ) from exc


def _measure_from_coverage_db(db_path: Path) -> Optional[float]:
    """Fallback: read percentage straight from coverage.py's SQLite file."""
    try:
        import coverage  # type: ignore
    except ImportError:
        return None
    if not db_path.exists():
        return None
    cov = coverage.Coverage(data_file=str(db_path))
    cov.load()
    return cov.report(show_missing=False, file=sys.stderr)


def main(argv: Optional[list[str]] = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--coverage",
        default=str(DEFAULT_COVERAGE_XML),
        help="Path to coverage.xml (Cobertura format). Default: backend/coverage.xml",
    )
    parser.add_argument(
        "--baseline-file",
        default=str(DEFAULT_BASELINE_FILE),
        help="Path to the committed baseline file. Default: backend/.coverage-baseline",
    )
    parser.add_argument(
        "--baseline",
        type=float,
        default=None,
        help="Override the baseline percentage explicitly (overrides --baseline-file).",
    )
    parser.add_argument(
        "--tolerance",
        type=float,
        default=DEFAULT_TOLERANCE_PP,
        help=f"Allowed drop below baseline, in percentage points. Default: {DEFAULT_TOLERANCE_PP}",
    )
    parser.add_argument(
        "--target",
        type=float,
        default=DEFAULT_TARGET_PERCENT,
        help=f"Aspirational target (informational only). Default: {DEFAULT_TARGET_PERCENT}",
    )
    parser.add_argument(
        "--write-baseline",
        action="store_true",
        help="When current coverage exceeds the baseline, rewrite the baseline file.",
    )
    args = parser.parse_args(argv)

    xml_path = Path(args.coverage)
    baseline_file = Path(args.baseline_file)

    baseline = args.baseline if args.baseline is not None else _read_baseline(baseline_file)
    measured = _measure_from_xml(xml_path)

    status = "OK"
    floor = baseline - args.tolerance
    dropped = measured + 1e-9 < floor

    # Pretty summary — one line per field so CI log search is easy.
    print("=" * 60)
    print("Phase 0 #7 — backend coverage gate")
    print("-" * 60)
    print(f"measured : {measured:6.2f} %")
    print(f"baseline : {baseline:6.2f} %  (from {baseline_file.name})")
    print(f"floor    : {floor:6.2f} %  (baseline - tolerance {args.tolerance}pp)")
    print(f"target   : {args.target:6.2f} %  (audit plan north star)")
    print("=" * 60)

    if dropped:
        status = "FAIL"
        print(
            f"[check_coverage] FAIL — coverage {measured:.2f}% dropped below "
            f"the floor {floor:.2f}% (baseline {baseline:.2f}%)."
        )
        print(
            "  → Either add tests to restore coverage, or if the drop is "
            "intentional (e.g. large refactor removed dead code), explicitly "
            "lower the baseline in a reviewable commit."
        )
        return 1

    if args.write_baseline and measured > baseline + args.tolerance:
        # Ratchet: commit the improved floor so future PRs cannot slide back.
        new_baseline = round(measured - args.tolerance, 2)
        baseline_file.write_text(f"{new_baseline:.2f}\n", encoding="utf-8")
        print(
            f"[check_coverage] ratcheted baseline {baseline:.2f}% → "
            f"{new_baseline:.2f}% in {baseline_file.name}"
        )

    if measured + 1e-9 < args.target:
        gap = args.target - measured
        print(
            f"[check_coverage] {status} — {gap:.2f}pp below the audit-plan "
            f"target of {args.target:.2f}%."
        )
    else:
        print(f"[check_coverage] {status} — target {args.target:.2f}% met or exceeded.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hostflow-frontend/scripts/check-coverage.mjs
#!/usr/bin/env node
/**
 * Phase 0 #7: frontend coverage ratchet gate.
 *
 * Consumes `coverage/coverage-summary.json` (emitted by the
 * `json-summary` vitest reporter — see `vitest.config.ts`) and enforces:
 *
 *   measured ≥ baseline - tolerance
 *
 * where `baseline` lives in `.coverage-baseline` (single float, 0-100).
 * The audit-plan target is 40% — currently aspirational; the script prints
 * the gap so CI readers see both the hard floor and the north star.
 *
 * Usage:
 *   npm run coverage:check                               # default paths
 *   node scripts/check-coverage.mjs --target 45 --write  # ratchet upward
 */
import { readFileSync, writeFileSync, existsSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, resolve } from 'node:path'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const FRONTEND_ROOT = resolve(__dirname, '..')

const DEFAULTS = {
  summary: resolve(FRONTEND_ROOT, 'coverage/coverage-summary.json'),
  baselineFile: resolve(FRONTEND_ROOT, '.coverage-baseline'),
  tolerance: 0.5,
  target: 40.0,
}

function parseArgs(argv) {
  const out = { ...DEFAULTS, write: false, overrideBaseline: null }
  for (let i = 2; i < argv.length; i += 1) {
    const flag = argv[i]
    const next = argv[i + 1]
    switch (flag) {
      case '--summary':
        out.summary = resolve(next)
        i += 1
        break
      case '--baseline-file':
        out.baselineFile = resolve(next)
        i += 1
        break
      case '--baseline':
        out.overrideBaseline = Number.parseFloat(next)
        i += 1
        break
      case '--tolerance':
        out.tolerance = Number.parseFloat(next)
        i += 1
        break
      case '--target':
        out.target = Number.parseFloat(next)
        i += 1
        break
      case '--write':
      case '--write-baseline':
        out.write = true
        break
      default:
        if (flag.startsWith('--')) {
          console.error(`[check-coverage] unknown flag: ${flag}`)
          process.exit(2)
        }
    }
  }
  return out
}

function readBaseline(path) {
  if (!existsSync(path)) {
    console.error(`[check-coverage] baseline file not found: ${path}`)
    process.exit(2)
  }
  const raw = readFileSync(path, 'utf-8').trim()
  if (!raw) {
    console.error(`[check-coverage] baseline file is empty: ${path}`)
    process.exit(2)
  }
  const value = Number.parseFloat(raw)
  if (Number.isNaN(value)) {
    console.error(
      `[check-coverage] baseline file must contain a single float (got "${raw}"): ${path}`,
    )
    process.exit(2)
  }
  return value
}

function readMeasured(summaryPath) {
  if (!existsSync(summaryPath)) {
    console.error(
      `[check-coverage] coverage summary not found: ${summaryPath}\n` +
        'Run `npm run test:coverage` first (emits coverage/coverage-summary.json).',
    )
    process.exit(2)
  }
  const raw = readFileSync(summaryPath, 'utf-8')
  let doc
  try {
    doc = JSON.parse(raw)
  } catch (err) {
    console.error(`[check-coverage] malformed JSON in ${summaryPath}: ${err.message}`)
    process.exit(2)
  }
  const pct = doc?.total?.lines?.pct
  if (typeof pct !== 'number') {
    console.error(
      `[check-coverage] could not read total.lines.pct from ${summaryPath}`,
    )
    process.exit(2)
  }
  return pct
}

function main() {
  const args = parseArgs(process.argv)
  const baseline =
    args.overrideBaseline != null ? args.overrideBaseline : readBaseline(args.baselineFile)
  const measured = readMeasured(args.summary)
  const floor = baseline - args.tolerance
  const dropped = measured + 1e-9 < floor

  const pad = (n) => n.toFixed(2).padStart(6)
  console.log('='.repeat(60))
  console.log('Phase 0 #7 — frontend coverage gate')
  console.log('-'.repeat(60))
  console.log(`measured : ${pad(measured)} %`)
  console.log(
    `baseline : ${pad(baseline)} %  (from ${args.baselineFile.replace(FRONTEND_ROOT + '/', '')})`,
  )
  console.log(`floor    : ${pad(floor)} %  (baseline - tolerance ${args.tolerance}pp)`)
  console.log(`target   : ${pad(args.target)} %  (audit plan north star)`)
  console.log('='.repeat(60))

  if (dropped) {
    console.error(
      `[check-coverage] FAIL — coverage ${measured.toFixed(2)}% dropped below ` +
        `the floor ${floor.toFixed(2)}% (baseline ${baseline.toFixed(2)}%).`,
    )
    console.error(
      '  → Either add tests to restore coverage, or if the drop is ' +
        'intentional, explicitly lower the baseline in a reviewable commit.',
    )
    process.exit(1)
  }

  if (args.write && measured > baseline + args.tolerance) {
    const newBaseline = Number((measured - args.tolerance).toFixed(2))
    writeFileSync(args.baselineFile, `${newBaseline.toFixed(2)}\n`, 'utf-8')
    console.log(
      `[check-coverage] ratcheted baseline ${baseline.toFixed(2)}% → ` +
        `${newBaseline.toFixed(2)}%`,
    )
  }

  if (measured + 1e-9 < args.target) {
    const gap = (args.target - measured).toFixed(2)
    console.log(
      `[check-coverage] OK — ${gap}pp below the audit-plan target of ${args.target.toFixed(2)}%.`,
    )
  } else {
    console.log(
      `[check-coverage] OK — target ${args.target.toFixed(2)}% met or exceeded.`,
    )
  }
}

main()
